internal/server/server.go

```diff
@@ -108,19 +176,104 @@ func (s *Server) handleLogin() http.HandlerFunc {
+		// Then in handleCallback() check the state in URL matches the session.Values["state"].
+		// However, our app does not do the full login -> redirect -> callback loop.
+		// The user hits /login, copies code, goes back to root page, and hits /callback.
+		// So there is no way to pass a state in the URL between the login and callback.
+		randomStr, err := random.SecureString(32)
 		if err != nil {
-			logAndError(w, http.StatusInternalServerError, err, "error generating random string")
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for state")
 			return
 		}
 
-		session.Values["rand"] = rs
+		// generate nonce
+		nonce, err := random.SecureString(32)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "error generating random string for nonce")
+			return
+		}
+		session.Values["nonce"] = nonce
 
 		if err := session.Save(r, w); err != nil {
 			logAndError(w, http.StatusInternalServerError, err, "error saving session")
 			return
 		}
 
-		fmt.Fprintf(w, "session saved")
+		http.Redirect(w, r, s.oauth2Config.AuthCodeURL(randomStr, oidc.Nonce(nonce)), http.StatusSeeOther)
+	}
+}
+
+func (s *Server) handleCallback() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// handle requests that contain an error
+		if err := r.URL.Query().Get("error"); err != "" {
+			logAndError(w, http.StatusBadRequest, fmt.Errorf(err+": "+r.URL.Query().Get("error_description")), "error in request")
+			return
+		}
+
+		session := s.getSession(r)
+		oidcContext := oidc.ClientContext(r.Context(), s.client)
+
+		// handle requests without an authorization code
+		code := r.URL.Query().Get("code")
+		if code == "" {
+			logAndError(w, http.StatusUnauthorized, fmt.Errorf("authorization code empty"), "error in authorization code")
+			return
+		}
+
+		// convert authorization code into an OAuth2 token
+		oauth2Token, err := s.oauth2Config.Exchange(oidcContext, code)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error converting code to token")
+			return
+		}
+
+		// extract the ID token from OAuth2 token
+		rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+		if !ok {
+			logAndError(w, http.StatusInternalServerError, fmt.Errorf("error extracting id_token"), "no id_token in token response")
+			return
+		}
+
+		// verify the ID token
+		idToken, err := s.verifier.Verify(oidcContext, rawIDToken)
+		if err != nil {
+			logAndError(w, http.StatusInternalServerError, err, "failed to verify id_token")
+			return
+		}
+
+		// verify the OIDC nonce matches
+		if idToken.Nonce != session.Values["nonce"] {
+			logAndError(w, http.StatusUnauthorized, fmt.Errorf("error invalid nonce"), "id token nonce does not match session nonce")
+			return
+		}
+
+		// extract custom claims
+		var claims struct {
+			Email string `json:"email"`
+		}
+		if err := idToken.Claims(&claims); err != nil {
```

Comment:
@dskatz I'm currently stuck on mocking this call to Claims.

I know you mentioned if Exchange on L226 returns a proper oauth2Token then we don't need to mock Verify on L240 and Claims here on L256.
I tried that as shown in the commented out code in mocks.go L42.
But the code will panic on Verify on L240, because the keySet field is nil
https://github.com/coreos/go-oidc/blob/v3.0.0/oidc/verify.go#L305

The IDTokenVerifier comes from the NewProvider call on L128 and Verifier on L133.
I cannot set the remoteKeySet field in my mock of NewProvider because it is private

https://github.com/coreos/go-oidc/blob/v3.0.0/oidc/oidc.go#L99
```
// Provider represents an OpenID Connect server's configuration.
type Provider struct {
	issuer      string
	authURL     string
	tokenURL    string
	userInfoURL string
	algorithms  []string

	// Raw claims returned by the server.
	rawClaims []byte

	remoteKeySet KeySet
}
```

Thus IDTokenVerifier's keySet field is nil, and we panic.

I don't know how to resolve from here.